File: app/src/main/java/me/yxp/qfun/utils/qq/QQCurrentEnv.java
package me.yxp.qfun.utils.qq;

import android.app.Activity;

import java.lang.reflect.Field;
import java.util.Map;

import me.yxp.qfun.utils.hook.xpcompat.XposedHelpers;
import me.yxp.qfun.utils.reflect.ClassUtils;

public class QQCurrentEnv {
    private static String sCurrentUin;
    private static Object sKernelMsgService;
    private static Object sQQAppInterface;

    public static Activity getActivity() {
        try {
            Class<?> activityThreadClass = Class.forName("android.app.ActivityThread");
            Object activityThread = activityThreadClass.getMethod("currentActivityThread").invoke(null);
            Field activitiesField = activityThreadClass.getDeclaredField("mActivities");
            activitiesField.setAccessible(true);

            Map<?, ?> activities = (Map<?, ?>) activitiesField.get(activityThread);
            for (Object activityRecord : activities.values()) {
                Class<?> activityRecordClass = activityRecord.getClass();
                Field pausedField = activityRecordClass.getDeclaredField("paused");
                pausedField.setAccessible(true);

                if (!pausedField.getBoolean(activityRecord)) {
                    Field activityField = activityRecordClass.getDeclaredField("activity");
                    activityField.setAccessible(true);
                    return (Activity) activityField.get(activityRecord);
                }
            }
        } catch (Exception ignored) {
        }
        return null;
    }

    public static Object getKernelMsgservice() {
        if (sKernelMsgService == null) setKernelMsgService();
        return sKernelMsgService;
    }

    private static void setCurrentUin() {
        try {
            sCurrentUin = (String) XposedHelpers.callMethod(sQQAppInterface, "getCurrentAccountUin");
        } catch (Exception ignored) {
        }
    }

    private static void setKernelMsgService() {
        try {
            Class<?> iKernelIService = ClassUtils.load("com.tencent.qqnt.kernel.api.IKernelService");
            Object kernelService = XposedHelpers.callMethod(sQQAppInterface, "getRuntimeService", iKernelIService, "");
            Object msgService = XposedHelpers.callMethod(kernelService, "getMsgService");
            sKernelMsgService = XposedHelpers.callMethod(msgService, "getService");
        } catch (Throwable ignored) {
        }
    }

    public static String getCurrentUin() {
        return (sCurrentUin == null || sCurrentUin.isEmpty()) ? "global" : sCurrentUin;
    }

    public static String getCurrentNickname() {
        try {
            return (String) XposedHelpers.callMethod(sQQAppInterface, "getCurrentNickname");
        } catch (Exception e) {
            return "";
        }
    }

    public static Object getQQAppInterface() {
        return sQQAppInterface;
    }

    public static void setCurrentEnv(Object qqAppInterface) {
        sQQAppInterface = qqAppInterface;
        setCurrentUin();
        setKernelMsgService();
    }
}

File: app/src/main/java/me/yxp/qfun/utils/qq/QQCurrentEnv.kt
package me.yxp.qfun.utils.qq

import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import com.tencent.mobileqq.app.BusinessHandler
import com.tencent.mobileqq.app.QQAppInterface
import com.tencent.mobileqq.qroute.QRoute
import com.tencent.mobileqq.qroute.QRouteApi
import com.tencent.qqnt.kernel.api.IKernelService
import com.tencent.qqnt.kernel.api.impl.KernelServiceImpl
import me.yxp.qfun.utils.reflect.callStaticMethod
import me.yxp.qfun.utils.reflect.getObject
import me.yxp.qfun.utils.reflect.toClass
import mqq.app.MobileQQ
import mqq.app.api.IRuntimeService

@Suppress("DEPRECATION")
object QQCurrentEnv {

    val globalPreference: SharedPreferences by lazy {
        HostInfo.hostContext.getSharedPreferences(
            "QFun_Config_global",
            Context.MODE_MULTI_PROCESS
        )
    }

    val qQAppInterface
        get() = MobileQQ.getMobileQQ().peekAppRuntime() as QQAppInterface

    val activity: Activity?
        get() = runCatching {

            val activityThreadClass = "android.app.ActivityThread".toClass
            val activityThread = activityThreadClass.callStaticMethod("currentActivityThread")
            val activities =
                activityThread?.getObject("mActivities") as? Map<*, *>
                    ?: return@runCatching null

            activities.values.firstNotNullOfOrNull { record ->
                record ?: return@firstNotNullOfOrNull null

                val paused = record.getObject("paused") as Boolean
                if (!paused) {
                    record.getObject("activity") as Activity
                } else {
                    null
                }
            }
        }.getOrNull()


    val currentDir: String
        get() = "${HostInfo.moduleDataPath}$currentUin/"

    val currentUin: String
        get() = runCatching {
            qQAppInterface.currentUin
        }.getOrElse {
            globalPreference.getString("currentUin", "global")!!
        }

    val currentNickName: String?
        get() = qQAppInterface.currentNickname

    val kernelMsgService
        get() = runCatching {

            val kernelService = runtime<IKernelService>() as KernelServiceImpl
            kernelService.msgService.service

        }.getOrNull()

}


internal inline fun <reified T : QRouteApi> api(): T {
    return QRoute.api(T::class.java)
}

internal inline fun <reified T : IRuntimeService> runtime(): T {
    return QQCurrentEnv.qQAppInterface.getRuntimeService(T::class.java, "")
}

internal inline fun <reified T : BusinessHandler> handler(): BusinessHandler {
    return QQCurrentEnv.qQAppInterface.getBusinessHandler(T::class.java.name)
}